presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AddExchangesBelowPartialAggregationOverGroupIdRuleSet.java

```diff
@@ -0,0 +1,373 @@
+ *       - Partial Aggregation
+ *         - GroupId
+ *           - LocalExchange
+ *             - RemoteExchange
+ * </pre>
+ * <p>
+ * Rationale: GroupId increases number of rows (number of times equal to number of grouping sets) and then
+ * partial aggregation reduces number of rows. However, under certain conditions, exchanging the rows before
+ * GroupId (before multiplication) makes partial aggregation more effective, resulting in less data being
+ * exchanged afterwards.
+ */
+public class AddExchangesBelowPartialAggregationOverGroupIdRuleSet
+{
+    private static final Capture<ProjectNode> PROJECTION = newCapture();
+    private static final Capture<AggregationNode> AGGREGATION = newCapture();
+    private static final Capture<GroupIdNode> GROUP_ID = newCapture();
+    private static final Capture<ExchangeNode> REMOTE_EXCHANGE = newCapture();
+
+    private static final Pattern<ExchangeNode> WITH_PROJECTION =
+            // If there was no exchange here, adding new exchanges could break property derivations logic of AddExchanges, AddLocalExchanges
+            typeOf(ExchangeNode.class)
+                    .matching(e -> e.getScope().isRemote()).capturedAs(REMOTE_EXCHANGE)
+                    .with(source().matching(
+                            // PushPartialAggregationThroughExchange adds a projection. However, it can be removed if RemoveRedundantIdentityProjections is run in the mean-time.
+                            typeOf(ProjectNode.class).capturedAs(PROJECTION)
+                                    .with(source().matching(
+                                            typeOf(AggregationNode.class).capturedAs(AGGREGATION)
+                                                    .with(step().equalTo(AggregationNode.Step.PARTIAL))
+                                                    .with(nonEmpty(groupingColumns()))
+                                                    .with(source().matching(
+                                                            typeOf(GroupIdNode.class).capturedAs(GROUP_ID)))))));
+
+    private static final Pattern<ExchangeNode> WITHOUT_PROJECTION =
+            // If there was no exchange here, adding new exchanges could break property derivations logic of AddExchanges, AddLocalExchanges
+            typeOf(ExchangeNode.class)
+                    .matching(e -> e.getScope().isRemote()).capturedAs(REMOTE_EXCHANGE)
+                    .with(source().matching(
+                            typeOf(AggregationNode.class).capturedAs(AGGREGATION)
+                                    .with(step().equalTo(AggregationNode.Step.PARTIAL))
+                                    .with(nonEmpty(groupingColumns()))
+                                    .with(source().matching(
+                                            typeOf(GroupIdNode.class).capturedAs(GROUP_ID)))));
+
+    private static final double GROUPING_SETS_SYMBOL_REQUIRED_FREQUENCY = 0.5;
+    private static final double ANTI_SKEWNESS_MARGIN = 3;
```

Comment:
Was there any experimentation with these parameters?

Reply:
No, I just lef them as-is while porting the change over from https://github.com/trinodb/trino/pull/105. Test's for this weren't added either, and I could not come up with a good integ test to experiment with these